### DISCUSSION TITLE
Erroneous "SAWarning: nested transaction already deassociated from connection"?

### DISCUSSION THREAD

**Discussion by cjw296** [sqlalchemy/sqlalchemy#8621] [Usage Questions] [answered]
I'm unclear as to whether this is a bug or I'm doing something wrong so thought I'd ask here.
The following snippet is a "smallest reproducer" extracted from some [refactoring I'm doing](https://github.com/Mortar/mortar_mixins/pull/3); the intention is to run tests in transactions, with a top level transaction for DDL execution and then each test being run in a subtransaction to give clean database for each test:

```python
from os import environ

from sqlalchemy import create_engine, Column, Integer, CheckConstraint
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import declarative_base, Session

Base = declarative_base()

class Model(Base):
    __tablename__ = 'model'
    __table_args__ = (
        CheckConstraint('id < 0'),
    )
    id = Column(Integer, primary_key=True)


engine = create_engine(environ['DB_URL'], future=True)
conn = engine.connect()
transaction = conn.begin()
try:
    subtransaction = conn.begin_nested()
    try:
        Base.metadata.create_all(bind=conn, checkfirst=False)
        session = Session(conn)
        session.add(Model(id=1))
        try:
            session.flush()
        except IntegrityError:
            pass
    finally:
        subtransaction.rollback()
finally:
    transaction.rollback()
```

When run with `DB_URL` pointing at an empty postgres 14 database and using SQLAlchemy 1.4.41, this gives:

```bash
$ python reproducer.py 
/Users/chris/vcs/git/mortar_mixins/reproducer.py:36: SAWarning: nested transaction already deassociated from connection
  subtransaction.rollback()
```

The key appears to be the `CheckConstraint` on the table, if I comment that out then I don't get the warning.
I'm guessing that means it might be a bug?

**Comment by zzzeek** [comment, marked as answer]
well the warning is because an error is being generated, and the flush() process is actually rolling back that savepoint already.

I looked and this is currently expected.   it's why you see in the [example](https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites) there's a check for "is_active" to avoid this warning.   so you could do this to resolve:

```py
transaction = conn.begin()
try:
    subtransaction = conn.begin_nested()
    try:
        Base.metadata.create_all(bind=conn, checkfirst=False)
        session = Session(conn)
        session.add(Model(id=1))
        try:
            session.flush()
        except IntegrityError:
            pass
    finally:
        if subtransaction.is_active:
            subtransaction.rollback()
finally:
    transaction.rollback()
```

now that is ugly sure.  so what you would do to use more of the "built in" management is to use the context managers, most directly like this:

```py
transaction = conn.begin()
try:
    with conn.begin_nested():
        Base.metadata.create_all(bind=conn, checkfirst=False)
        session = Session(conn)
        session.add(Model(id=1))
        try:
            session.flush()
        except IntegrityError:
            pass
finally:
    transaction.rollback()
```

or the Session can do it

```py
transaction = conn.begin()
try:
    with Session(conn) as session:
        with session.begin_nested():
            Base.metadata.create_all(bind=session.connection(), checkfirst=False)
            session.add(Model(id=1))
            try:
                session.flush()
            except IntegrityError:
                pass
finally:
    transaction.rollback()
```

the context managers are testing if rollback() is needed when an exception occurs.

**Reply by cjw296** [reply]
Thanks!

```python
transaction = conn.begin()
try:
    with conn.begin_nested():
        ...
finally:
    transaction.rollback()
```
...doesn't quite do what I want: these are tests so I _always_ need to rollback.

The `if subtransaction.is_active:` version works, and no warning, _but_, still confused as to why it's needed:

- Why does this warning only occur when the `CheckConstraint` is present on the table?
- Why is `.flush()` rolling back the savepoint? That feels like something that should only be done by a context manager or an explicit rollback...

**Reply by zzzeek** [reply]
>     * Why does this warning only occur when the `CheckConstraint` is present on the table?

well you are only getting an exception throw when that constraint is present

> 
>     * Why is `.flush()` rolling back the savepoint? That feels like something that should only be done by a context manager or an explicit rollback...

cue the music!  https://docs.sqlalchemy.org/en/14/faq/sessions.html#but-why-does-flush-insist-on-issuing-a-rollback

**Reply by cjw296** [reply]
> > ```
> > * Why does this warning only occur when the `CheckConstraint` is present on the table?
> > ```
> 
> 
> well you are only getting an exception throw when that constraint is present

If I remove the `CheckConstraint` and replace `session.add(Model(id=1))` with `session.add(Model())` (ie: no primary key specified) and flush, I still get an `IntegrityError` but no warning...

**Reply by CaselIT** [reply]
> Why does this warning only occur when the CheckConstraint is present on the table?

the check constraint makes the flush fail. if it's not present the model it's inserted correctly

**Reply by cjw296** [reply]
@CaselIT - please see my follow up: If I remove the `CheckConstraint` and replace `session.add(Model(id=1))` with `session.add(Model())` (ie: no primary key specified) and flush, I still get an `IntegrityError` but no warning...

**Reply by CaselIT** [reply]
Primary key implies autoincrement so the integrity error is strange for the model above.
Without some additional information, like the db in use and the stack trace it's hard to know for sure if it's the flush that raises the error

**Reply by zzzeek** [reply]
the example runs, let's try

**Reply by zzzeek** [reply]
if I take the CHECK constraint out, and put a "raise" under that IntegrityError, whether or not I create Model(id=1) or Model() (the latter means it will use SERIAL to autoincrement), the program completes with no errors as expected, the "raise" is not reached.  If I restore the CHECK constraint, the raise is reached:

```
sqlalchemy.exc.IntegrityError: (psycopg2.errors.CheckViolation) new row for relation "model" violates check constraint "model_id_check"
DETAIL:  Failing row contains (1).

[SQL: INSERT INTO model (id) VALUES (%(id)s)]
[parameters: {'id': 1}]
```

**Reply by cjw296** [reply]
Yeah, apologies, must have missed something when I attempted this before, the following does, indeed, give the warning:

```python
Base = declarative_base()

class Model(Base):
    __tablename__ = 'model'
    id = Column(Integer, primary_key=True)
    other = Column(Integer, nullable=False)


engine = create_engine(environ['DB_URL'], future=True)
conn = engine.connect()
transaction = conn.begin()
try:
    subtransaction = conn.begin_nested()
    try:
        Base.metadata.create_all(bind=conn, checkfirst=False)
        session = Session(conn)
        session.add(Model(id=1))
        try:
            session.flush()
        except IntegrityError:
            pass
    finally:
        subtransaction.rollback()
finally:
    transaction.rollback()
```

**Comment by cjw296** [comment]
Newer version of hitting this error:

https://github.com/sqlalchemy/sqlalchemy/discussions/12006
